server/sonar-db-dao/src/main/java/org/sonar/db/measure/MeasureQuery.java

```diff
@@ -0,0 +1,242 @@
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.db.measure;
+
+import java.util.Collection;
+import java.util.Objects;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Collections.singleton;
+import static java.util.Objects.requireNonNull;
+
+public class MeasureQuery {
+  @CheckForNull
+  private final String analysisUuid;
+
+  @CheckForNull
+  private final Collection<String> projectUuids;
+
+  @CheckForNull
+  private final Collection<String> componentUuids;
+
+  @CheckForNull
+  private final Collection<Integer> metricIds;
+
+  @CheckForNull
+  private final Collection<String> metricKeys;
+
+  @CheckForNull
+  private final Long personId;
+
+  private MeasureQuery(Builder builder) {
+    this(builder.analysisUuid, builder.projectUuids, builder.componentUuids, builder.metricIds, builder.metricKeys, builder.personId);
+  }
+
+  private MeasureQuery(@Nullable String analysisUuid,
+    @Nullable Collection<String> projectUuids,
+    @Nullable Collection<String> componentUuids,
+    @Nullable Collection<Integer> metricIds,
+    @Nullable Collection<String> metricKeys,
+    @Nullable Long personId) {
+    checkArgument(metricIds == null || metricKeys == null, "Metric IDs and keys must not be set both");
+    checkArgument(projectUuids != null || componentUuids != null, "At least one filter on component UUID is expected");
+    checkArgument(componentUuids == null || componentUuids.size() == 1 || (projectUuids != null && projectUuids.size() == 1),
+      "Component UUIDs can only be used when a single project UUID is set");
+
+    this.analysisUuid = analysisUuid;
+    this.projectUuids = projectUuids;
+    this.componentUuids = componentUuids;
+    this.metricIds = metricIds;
+    this.metricKeys = metricKeys;
+    this.personId = personId;
+  }
+
+  public String getAnalysisUuid() {
+    return analysisUuid;
+  }
+
+  @CheckForNull
+  public Collection<String> getProjectUuids() {
+    return projectUuids;
+  }
+
+  @CheckForNull
+  public String getProjectUuid() {
+    return isOnComponents() ? projectUuids.iterator().next() : null;
+  }
+
+  @CheckForNull
+  public Collection<String> getComponentUuids() {
+    return componentUuids;
+  }
+
+  @CheckForNull
+  public String getComponentUuid() {
+    return isOnSingleComponent() ? componentUuids.iterator().next() : null;
+  }
+
+  @CheckForNull
+  public Collection<Integer> getMetricIds() {
+    return metricIds;
+  }
+
+  @CheckForNull
+  public Collection<String> getMetricKeys() {
+    return metricKeys;
+  }
+
+  @CheckForNull
+  public Long getPersonId() {
+    return personId;
+  }
+
+  public boolean returnsEmpty() {
+    return (projectUuids != null && projectUuids.isEmpty())
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Reduce the number of conditional operators (7) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)